util/Pos/Util/Verification.hs

```diff
@@ -0,0 +1,96 @@
+{- |
+
+Machinery to support pure datatype verification.
+
+1. Datatypes can have and not have internal pure consistency. If
+datatype doesn't have any implicit rules, it's called safe. 'Integer'
+is safe, 'Either Text String' is safe too. Otherwise, it's called
+unsafe. 'Coin' is unsafe, because it has implicit upper bound on the
+value, which is lower than 'Word64' bound. Newtype over string that is
+supposed to contain only strings of length less than 30 symbols is
+unsafe.
+
+2. Unsafe types' constructors should have prefix \"Unsafe\" (for
+example, 'UnsafeProxySecretKey'). They should also have description of
+why are they unsafe and provide some easy-to-use builders (@mkX@) if
+needed.
+
+3. If datatype @S@ is a part of @T@ (e.g. record field), then @T@
+is not called unsafe.
+
+4. @T@ is called verifiable if it is unsafe or any of its parts are
+verifiable.
+
+5. Every verifiable type must have 'PVerifiable' instance. See
+description of this class.
+
+-}
+module Pos.Util.Verification
+    (
+      PVer
+    , pverFail
+    , pverField
+
+    , PVerifiable(..)
+    , runPVerify
+    , runPVerifyFail
+    , runPVerifyPanic
+    ) where
+
+import           Universum
+
+import qualified Data.Text as T
+
+import           Pos.Util.Util (eitherToThrow)
+
+-- Prefix P is annoying, but we need to distinguish between pure
+-- verification and VAR somehow.
+
+-- | Pure verification monad. It accumulates errors along the
+-- verification (AST) path.
+newtype PVer a = PVer
+    { getPVer :: Either [Text] a
+    } deriving (Show, Eq, Functor, Applicative, Monad)
+
+data VerError = VerError Text deriving Show
+
+instance Exception VerError
+
+-- | Fail inside the 'PVer' monad.
+pverFail :: Text -> PVer ()
+pverFail t = PVer (Left [t])
+
+-- | Verifies some field, prefixing with the text value in case of
+-- error. Prefix is supposed to be the record field name. Use it when
+-- you want to specify which field/component are you verifying.
+pverField :: Text -> PVer () -> PVer ()
+pverField p (PVer v) = PVer $ first (\x -> p:x) v
```

Comment:
Hm, so `Left` in `getPVer` accumulates values which are semantically different, right? It looks like a bad idea :/

Reply:
Do we have an idea how to modify `PVer` type to do proper exception handling (using ADTs instead of `Text`)?

Reply:
@gromakovsky I do not quite understand what does "semantically different" mean here. What I wanted to express with `PVer` is that user can provide hints along the verification path. The good analogy is "the term we failed to parse". It cannot be simply (without TH) automated. I think it's more or less optimal solution. Number of places this is used in is low. It is mostly needed when `A` has two fields of type `B` and we want to point out which one lead to error.

>Do we have an idea how to modify PVer type to do proper exception handling

I think we could define a type family. But again, despite agreeing this is important, I consider this to be a separate task.